common/src/reserved.rs

```diff
@@ -22,55 +22,124 @@ pub struct ReservedState {
-        for member in &self.members {
-            if let Some(delegatee) = &member.consensus_delegatee {
-                validator_set
-                    .entry(delegatee.clone())
-                    .and_modify(|v| *v += member.consensus_voting_power)
-                    .or_insert(member.consensus_voting_power);
-            } else {
-                validator_set
-                    .entry(member.name.clone())
-                    .and_modify(|v| *v += member.consensus_voting_power)
-                    .or_insert(member.consensus_voting_power);
-            }
-        }
-        // TODO handle error
-        Ok(self
-            .consensus_leader_order
+        let validator_set = self
+            .members
             .iter()
-            .map(|name| (self.query_public_key(name).unwrap(), validator_set[name]))
-            .collect())
+            .map(|member| {
+                let name = member
+                    .consensus_delegatee
+                    .as_ref()
+                    .map_or_else(|| member.name.as_str(), |name| name.as_str());
+                let public_key = self.query_public_key(&name.to_string()).ok_or_else(|| {
+                    format!("the public key of {name} is not found in the reserved state.")
+                })?;
+                Ok((public_key, member.consensus_voting_power))
+            })
+            .try_fold(
+                HashMap::new(),
+                |mut map, result: Result<(crypto::PublicKey, u64), String>| {
+                    let (public_key, voting_power) = result?;
+                    map.entry(public_key)
+                        .and_modify(|v| *v += voting_power)
+                        .or_insert(voting_power);
+                    Ok::<HashMap<crypto::PublicKey, u64>, String>(map)
+                },
+            )?
+            .into_iter()
+            .collect::<Vec<_>>();
+
+        Ok(validator_set)
     }
 
     pub fn get_governance_set(&self) -> Result<Vec<(PublicKey, VotingPower)>, String> {
-        let mut governance_set = HashMap::new();
-        for member in &self.members {
-            if let Some(delegatee) = &member.governance_delegatee {
-                governance_set
-                    .entry(delegatee.clone())
-                    .and_modify(|v| *v += member.consensus_voting_power)
-                    .or_insert(member.consensus_voting_power);
-            } else {
-                governance_set
-                    .entry(member.name.clone())
-                    .and_modify(|v| *v += member.consensus_voting_power)
-                    .or_insert(member.consensus_voting_power);
-            }
-        }
-        Ok(governance_set
+        let governance_set = self
+            .members
             .iter()
-            .map(|(name, voting_power)| (self.query_public_key(name).unwrap(), *voting_power))
-            .collect())
+            .map(|member| {
+                let delegatee_name = member
```

Comment:
```suggestion
                let name = member
```
The variable name is misleading and also, no consistency with `get_validator_set` so I suggest you to fix.

Reply:
okay. I would change it.